### DISCUSSION TITLE
How to change the default “Squash & Merge” commit message?

### DISCUSSION THREAD

**Discussion by arthurio** [community/community#16271] [Pull Requests]
Migrated from https://github.community/t/how-to-change-the-default-squash-merge-commit-message/1155/2

> When “Squash & Merge” a branch in a PR, the default commit message is either:
>   - the commit message of the the commit in the PR, if the PR has a single commit;
>   - the title of the PR’s and the description from all commits’ titles, if the PR has more than one commits.
> 
> As this behavior is inconsistent, it often result in inconsistent merge messages.  Is there a way to configure the default commit message for “Squash & Merge”?

Original author: https://github.community/u/ashi009

**Comment by queenofcorgis** [comment]
Hey @arthurio thanks for posting this 😄

**Comment by tolomea** [comment]
+1

Also would be nice to have an option to have it be the title and description from the pull request.

**Comment by spaltrowitz** [comment]
this has shipped! 🚢  [Default to PR titles for squash merge commit messages | GitHub Changelog](https://github.blog/changelog/2022-05-11-default-to-pr-titles-for-squash-merge-commit-messages/)

**Comment by arthurio** [comment]
@spaltrowitz I don't think that's quite enough... The change only takes the PR title as commit message. It doesn't put the description details in the commit message which is (for me) the most important.

**Reply by shmeadyy** [reply]
Hi @arthurio! Thanks for sharing that feedback, I agree that this isn't enough! We're looking into making some further changes around squash/merge commits, specifically. I'd love to know what else you're thinking about that could help? Are you looking for the PR description to also become the commit message? Anything else we can be thinking about?

**Comment by kenji-miyake** [comment]
@spaltrowitz Thank you for shipping it! It's very useful. :smile: 

But please let me request one thing. is it possible to support either of the following features?:
- Disable auto-merge after editing the PR title.
- Require canceling auto-merge to edit the title.

Currently, if I edit the PR title after auto-merge is enabled, the old title is used for the squashed commit message, I feel it's not desired behavior.

Here is an example.
https://github.com/autowarefoundation/autoware_msgs/pull/17#issuecomment-1121764886

**Reply by shmeadyy** [reply]
Hey @kenji-miyake, thanks for that feedback! When you select 'Default to PR title for squash merge commit messages' in the Repo settings, does the above still happen on auto-merge?

**Reply by kenji-miyake** [reply]
@shmeadyy Unfortunately yes, this is my test result.
https://github.com/kenji-miyake/test-auto-merge/pull/1

**Comment by hernaneg350** [comment]
I'm giving a huge plus 1 to @arthurio 's comments. As multiple other people have commented on https://github.community/t/how-to-change-the-default-squash-merge-commit-message/1155/62 (the original feature request in github community), when squashing, a lot of users are explicitly wanting to _discard_ the commit history. Squash is a great feature to allow a clean history in permanent branches, while allowing flexibility and change browsing in the PR. 

I think having a configuration that allows:
- The PR title to be the commit title
- The PR description to be the commit message

Would be an amazing improvement over what you have now. 

A further step would be the ability to have templates for the squash message (or even other merge options).

**Reply by binoche9** [reply]
+1 to also including the PR description - only using the PR title significantly limits the amount of context we can store in commit messages in our git history

**Reply by derekmurawsky** [reply]
Template would be excellent. We leverage Conventional Commits with a required body and would love to have that template put in place at squash time to not put burden on the devs making minor changes in a PR.

**Comment by christianpinheiro** [comment]
+1 thanks for help

**Comment by DylanGuedes** [comment]
Could you also consider adding an option to have the PR description in the merge commit message? As you can see [in this gigantique thread](https://github.community/t/how-to-change-the-default-squash-merge-commit-message/1155/66) that feature would be highly desirable by tons of people.

Especially for bigger projects with a lot of contributors, not having this is really impactful. Unless all maintainers pay 100% attention to this at every new merge, the commit history of the project ends up having a bunch of "Signed-off-by", "tmp", "lint fix" and other useless phrases everywhere. That feature would completely solve it, though.

**Reply by comp615** [reply]
I think I saw @shmeadyy had said they were working on this, but big +1. 

Right now git history or blame is not super useful on its own, including the PR description _(and a link back to the Github PR!)_ on squash would drastically speed up git spelunking expeditions by providing context in editors instead of having to grab the sha, search GH, find the PR, etc.

**Reply by DylanGuedes** [reply]
@spaltrowitz do you think it would be better if we instead proceed to open a new discussion just for this?

**Reply by zachariahcox** [reply]
+1 to adding an option to include a link to the gh pr in the commit message

**Comment by willsmythe** [comment]
The feature to let admins control the default merge message presented to users when they merge a pull request is in progress! Here's an early mock at this new repo-level setting:

![image](https://user-images.githubusercontent.com/2503052/181119357-028c17d4-ed37-4d8c-b526-6811cb617827.png)

**Reply by derekmurawsky** [reply]
Looks good. Looking forward to this feature!

**Reply by camilaagw** [reply]
@willsmythe has this been already released?

**Comment by keviniscodingai** [comment]
Would anyone happen to know? https://github.com/orgs/community/discussions/41462

**Comment by dhimmel** [comment]
My preference for squash merging is to use full PR urls. Like if this discussion were a PR, I'd want the squash commit to contain:

```
merges https://github.com/orgs/community/discussions/16271
```

and not

```
(#16271)
```

These relative issue/PR links cause problems when commits are ported to different repositories, like if a codebase is split or if a fork is not tracked properly.

So a custom Squash & Merge commit message would ideally support full PR urls.

**Comment by joevandyk** [comment]
It looks like https://github.com/orgs/community/discussions/16271#discussioncomment-3255724 is released and answers this question.

**Comment by karlismelderis** [comment]
I stumbled over similar issue.

Original our pain was maintaining clean commit messages in PR, read fixups and time lost to make it happen. We gave up 😎

Then squash on merge arrived and issue to summarize all commits and what's done arrived.
We failed, no one liked to do it again as PR title was already shaped 😎

Then we enabled to take PR body...
Everyone is happy but we pull in a lot of garbage that people don't want or forget to clean up.

Adding start and end tags that would mark what counts as commit message in PR body would be a big boon for us.
